### ISSUE TITLE
Request for source code (.py files)

### ISSUE THREAD

**Issue by xiaolan0195** [Yubo-Huang/maobc#1]
Hi @LuckyYubo,

I am very interested in your paper "Optimizing an offline reinforcement learning control policy for wind farms".

However, the current repository only contains `.pyc` files and no `.py`
 source code, making it impossible to run or study.

Could you please upload the complete source code (especially `bc_agent.py`, `agents.py`
, etc.)?

Thank you!

**Comment by Yubo-Huang** [comment]
Hi @xiaolan0195 Thanks for reaching out. I think this repo has already included the .py files. The reason you can not download maybe I changed my usename previously. 

Can you please use this command to download my repo?

```
git clone https://github.com/Yubo-Huang/maobc.git
```

or directly download the zip file from this webpage.

Best

**Comment by xiaolan0195** [comment]
> 嗨，感谢您的联系。我认为这个存储库已经包含了.py文件。您无法下载的原因可能是我之前更改了我的使用名。
> 
> 你能用这个命令下载我的存储库吗？
> 
> ```
> git clone https://github.com/Yubo-Huang/maobc.git
> ```
> 
> 或直接从此网页下载zip文件。
> 
> 最好

Thank you for your answer! Yes, I am from this website https://github.com/Yubo-Huang/maobc# After downloading the installation package, unpacking it and running it, in the agents.py file, it shows that the four libraries bc_agent, bcq.agent, brac_prims_agent, and sac.agent have not been imported. Then I searched for the files in the package, but I couldn't find these four files.
import bc_agent
import bcq_agent
import brac_dual_agent
import brac_primal_agent
import sac_agent

Thank you！

**Comment by xiaolan0195** [comment]
Hi @LuckyYubo,
 I want to generate training data for intelligent agents to prepare for their subsequent learning.What is the training data format for this intelligent agent? I noticed that there is no data explanation in the Excel file of the data folder. I would be extremely grateful if you could answer.
Thank you!

**Comment by Yubo-Huang** [comment]
Hi @xiaolan0195 

When you use maobc, in fact, you donot need the four files and you can directly delete these lines. if you would like to test these algorithms, you could find the source code from this repo: `https://github.com/google-research/google-research/tree/master/behavior_regularized_offline_rl`

For the dataset issue, I directly use csv files to save my training data, but they are too large to upload. The data format is [s, a, r, s']. For examples, if the dim of [s, a, r, s'] is [10, 3, 1, 10], and you have 100 samples, the array you save in you dataset is 100*(10+3+1+10), and then you can make you own pytroch or tensorflow dataset to read and batch your training samples. If you csv file is to large, I recommand to the HDF5 data format. 

Hope this is helpful,

Best

**Comment by xiaolan0195** [comment]
> 你好
> 
> 当你使用maobc的时候，其实不需要这四个文件，你可以直接删除这几行。如果您想测试这些算法，可以从此存储库中找到源代码：`https://github.com/google-research/google-research/tree/master/behavior_regularized_offline_rl`
> 
> 对于数据集问题，我直接使用 csv 文件来保存我的训练数据，但它们太大而无法上传。数据格式为 [s， a， r， s']。例如，如果 [s， a， r， s'] 的暗度为 [10， 3， 1， 10]，并且您有 100 个样本，则您保存在数据集中的数组为 100*（10+3+1+10），然后您可以制作自己的 pytroch 或 tensorflow 数据集来读取和批处理您的训练样本。如果您的 csv 文件太大，我建议使用 HDF5 数据格式。
> 
> 希望对您有所帮助，
> 
> 最好

Hi @LuckyYubo,
Thank you very much for your answer! I would also like to ask if the dataset can be obtained through the FLORIS series wind farm simulation software? And then train the agent through the training set? The data format is [s, a, r, s'], where s is the current state and s' is the new time state. How should we understand this data? For example, if the rotor speed is one of the states and simulation can obtain the rotor speed (s), how do we understand s'? My research direction is multi-agent collaborative control in wind farms. I have just started learning about it and still have some doubts about these details. I would be extremely grateful if you could answer.
Thank you!

**Comment by Yubo-Huang** [comment]
I used WFSim to generate the training datasets, which are dynamic. You can use FLORIS to produce your own data as well, but I’m not sure how well it will perform, since it is a static model.

Best

**Comment by xiaolan0195** [comment]
> 我用WFSim生成了动态训练数据集。你也可以用FLORIS生成自己的数据，但我不确定它的表现如何，因为它是静态模型。
> 
> 最好
Hello！
Thank you very much for your response! I am currently using dynamic simulation FAST. Farm and have generated 600s of data for 6 wind turbines. The state variables s are rotor speed, generator speed, and generator torque, while the control variables a are yaw angle and r are power. The data format of the table is [s, a, r, s'], which forms the training data table. s is the corresponding variable of rotor speed, generator speed, and generator torque in the out file of each wind turbine. a is the yaw angle corresponding to each wind turbine. r is the current power output of the wind turbine itself, or does this value take the power of the entire wind farm? If s' is the next time state, is it equivalent to taking the data of s state at the next time? For example, if the rotor speed has 10000 rows of data, moving forward relative to 9999 rows after taking the next time, wr '[1,1] corresponds to wr [2,1]? Since a is a fixed simulation run during the running process, setting different values of a and generating multiple data combinations together, is this the training data?
Best！

**Comment by Yubo-Huang** [comment]
Hi.Fast.Farm is an excellent tool for simulating wind farms. I believe that the reward should be tied to your control objective—whether that’s maximizing the overall wind farm power output or the output of individual turbines. I’m not familiar with your specific data format, but to my knowledge, the first 30 seconds of output data from Fast.Farm or OpenFAST should not be used, as this period corresponds to the turbine startup phase.

Best

**Comment by xiaolan0195** [comment]
> Hi.Fast.Farm 是一个极好的风电场模拟工具。我认为奖励应与你的控制目标挂钩——无论是最大化整体风电场的发电量，还是单个风机的输出。我不太熟悉你的具体数据格式，但据我所知，Fast.Farm或OpenFAST输出的前30秒数据不应使用，因为这段时间对应涡轮启动阶段。
> 
> 最好

Thanks a lot. I will try it out. May I ask for your email and other contact information? May I still ask you if I encounter any problems in the future?
Best!

**Comment by Yubo-Huang** [comment]
You can paste you question here, then I will reply you ASAP.

Best.

**Comment by xiaolan0195** [comment]
Hello！
I don't understand the setting of the reward function r given in the paper, as shown in Figure 1. May I ask if this has been calculated and placed in the CSV file for training? Or is the 'r' column in the CSV file for calculating the reward function in the code?
<img width="900" height="680" alt="Image" src="https://github.com/user-attachments/assets/140dff78-7597-4a81-b033-5fc96b3fb943" />
As shown in Figure 2, when I look at the 25 column record of power in the CSV file, it seems that this value is not calculated by the formula and does not correspond to it.
<img width="323" height="1221" alt="Image" src="https://github.com/user-attachments/assets/f8d4ff44-a8c9-4cc6-b213-398730cf6d9a" />
How should I understand this reward function r?
Best！
